Refactor/reduce complexity

NacerKROUDIR opened the pull request:
I removed the `pr_to_unp` parameter that lets the user select how to order the different demographic groups (from most privileged to most unprivileged or the reverse) for two reasons:
- It is easy for the user to reverse the order once we have it.
- It adds so much complexity to the code.

I also moved the initial computation of the differences or ratios to and independent function (`_compute`). There are two functions that rely on this method directly (`summary` and `rank`). This refactor optimises the code and avoid doing unnecessary competitions that were present in the previous implementation.

MarcBresson commented:
Follow PEP8:
- line break for long string (use implicit string continuation): wtf is that thing L.62 of metrics.py
- `Metric` should be an ABC
- all your classes should inherit your ABC `Metric`, (FairnessMetricDifference is a Metric, right?)
- make your code breath, don't be scared of blank lines
- stop using attributes everywhere since you return the value of the attribute anyway (see self.ranking)
- ConfusionMetrics.__call__ is awful to look at, refactor and improve. This is not clear what it does.

NacerKROUDIR commented:
- Applied line break for long strings.
- Added ABC abstract class and abstract method in `Metric`.
- `Metric`, `FairnessMetricDifference` and `FairnessMetricRatio` are different base classes that gets inherited by the actual metrics available for the user.
- `Metric`: is for performance evaluation for the different demographic groups.
- `FairnessMetricDifference`: is for fairness metrics that are based on difference for calculating disparity.
- `FairnessMetricRatio`: is for fairness metrics that are based on ratio for calculating disparity.
- Added line breaks for readability.
- I try to use attributes when I have to access the variable in another method. (Ex: `self.ranking` is returned by `.rank()` method and it is used in the `.is_biased()` method). In other places I do it because of `mypy` complaints.
- Refactored `ConfusionMatrix.__call__()` and added a comment to explain.

MarcBresson commented on fair_mango/metrics/metrics.py:
```diff
@@ -347,39 +362,37 @@ def __call__(self) -> tuple[Sequence, list]:
                     tp = len(real_values)
                     tn = fp = fn = 0
+
                 elif (np.unique(real_values) == 0).all() and (
                     np.unique(predicted_values) == 0
                 ).all():
                     tn = len(real_values)
                     tp = fp = fn = 0
-                elif (np.unique(real_values) == 1).all() and (
-                    np.unique(predicted_values) == 0
-                ).all():
-                    fn = len(real_values)
-                    fp = tn = tp = 0
-                elif (np.unique(real_values) == 0).all() and (
-                    np.unique(predicted_values) == 1
-                ).all():
-                    fp = len(real_values)
-                    fn = tn = tp = 0
+
                 else:
```
use `confusion_matrix(..., labels=[0, 1])` instead

MarcBresson commented:
https://github.com/datategy/Fair-Mango/pull/10/files#diff-f60460593beef4b2e9574ac4bd0492afb517eaac1129c634e527dd38109471fdL75-L78

refactor all these functions by using kwargs and silencing mypy

MarcBresson commented on fair_mango/metrics/metrics.py:
```diff
@@ -442,17 +461,21 @@ def __call__(self) -> tuple[Sequence, list]:
                     predicted_values = predicted_y_group
+
                 for metric_name, metric in self.metrics.items():
                     result_for_group.setdefault(metric_name, []).append(
```
if metric_name is "sensitive" it s gonna raise an error

NacerKROUDIR replied:
Good catch! I created a check that raises a `KeyError` in case one of the metrics has the label `"sensitive"`

MarcBresson commented on fair_mango/metrics/metrics.py:
```diff
@@ -442,17 +461,21 @@ def __call__(self) -> tuple[Sequence, list]:
         return self.data.real_target, self.results
 
 
 def difference(result_per_groups: np.array) -> dict:
```
the type is `np.ndarray`. Do not take my word for it, check with google

be more ambitious about your typing. Check if it returns `dict[str, list[float] | float]` or not

NacerKROUDIR replied:
You are right!

MarcBresson commented on fair_mango/metrics/metrics.py:
```diff
@@ -442,17 +461,21 @@ def __call__(self) -> tuple[Sequence, list]:
                         metric(real_values, predicted_values)
                     )
+
             self.results.append(result_for_group)
```
should not be in an attribute

MarcBresson commented on fair_mango/metrics/metrics.py:
```diff
@@ -480,10 +505,11 @@ def ratio(
             result[key] = result_i / result_j
```
you can divide by zero with numpy. It s gonna give you "inf":

```python
a = np.array(0)
b = np.array(1)
b / a
>>> inf
```

NacerKROUDIR replied:
Yup. I removed the `zero_division` parameter completely.

MarcBresson commented on fair_mango/metrics/metrics.py:
put ABCs in a separate file

NacerKROUDIR replied:
Good idea! a new file `fair_mango/metrics/base.py` is created and it includes all the base classes and functions that are not exposed to the user.
The classes:
- `Metric`
- `FairnessMetricDifference`
- `FairnessMetricRatio`
The functions:
- `is_binary`
- `encode_target`
- `false_negative_rate`
- `false_positive_rate`
- `true_negative_rate`
- `true_positive_rate`
- `difference`
- `ratio`

MarcBresson commented on fair_mango/metrics/metrics.py:
```diff
@@ -549,17 +586,20 @@ def summary(self) -> dict:
-        if self.result is None:
-            self.summary()
+    def rank(self) -> dict:
         result: dict = {}
```
be more ambitious about your typing

NacerKROUDIR replied:
Typing for the returned values is improved in many methods now.

MarcBresson commented on fair_mango/metrics/metrics.py:
```diff
@@ -549,17 +586,20 @@ def summary(self) -> dict:
+
         for key, values in self.results.items():
             if isinstance(values, float):
                 values = [values]
```
this check can be useless depending on the return type of `def difference(...) -> ...`. If `difference` returns only a dict of list, then you are never gonna enter this condition

NacerKROUDIR replied:
I checked the `.difference()` function and it returns `dict[tuple, float | np.ndarray[float]]. Which means that the check I am doing is needed.
PS: I added the typing to the returned value of `.difference()` and `.ratio()` functions.

MarcBresson replied:
in what case does it return a float ?

MarcBresson replied:
do you think it would be better to make difference() always output a numpy array to clean up higher level code?

NacerKROUDIR replied:
Done. I made sure the functions `difference` and `ratio` always return a 1D numpy array.

MarcBresson replied:
FYI, you use `MAINT` too often I think. `MAINT` is meant (ahah) for :

> MAINT: maintenance commit (refactoring, typos, etc.)

When you improve your code you can use `ENH` instead

MarcBresson replied:
ratio and difference are way too similar to be two different functions. You can try to use a generator to factorize

NacerKROUDIR replied:
Since the code does the same thing in another way I consider refactoring and use `MAINT` tag. I try to use `ENH` when I add something new.
Should I use `ENH` for refactoring commits that optimise the code too?

MarcBresson replied:
hummm good question, I was seing refactorings as non-functional changes, but in a sense you are right. The higher level interface did not change.
Well done

NacerKROUDIR commented on fair_mango/metrics/base.py:
```diff
@@ -212,7 +213,9 @@ def __init__(
-def difference(result_per_groups: np.ndarray) -> dict[tuple, np.ndarray[float]]:
+def calculate_disparity(
+    result_per_groups: np.ndarray, method: Literal["difference", "ratio"]
+) -> dict[tuple, np.ndarray[float]]:
```
I used generator to get the disparity values. However, I don't see the benefit since the operation is done once on a numpy array (which directly does it element by element). 

Also, we are dealing with one value or two for the most part so the memory usage difference will be negligible too.
